### PR TITLE
Fix Date Bug in Month View + Tooltip on Hover

### LINKED ISSUE #1590: [BUG] [Feature] [kalender] Eintrag Termin in Monatsansicht Zeiten krumm => ggfs. Mouseover für Zeit unter dem Cursor?
erwartetes Verhalten ist: halbstündige Anlage, d.h. zu :00 oder :30 
hier: Macbook pro 16 Zoll, Retina Display. Aber auch mit externem Display Widescreen beobachtet 

Vielleicht könnte es ja auch bei dem Verhalten bleiben und zur Hilfe/Unterstützung wird ein Mouseover eingeblendet, mit der Zeit unter dem Mauszeiger. Das das wäre der Königsweg! Denn so können minutengenaue Termine angelegt werden, ohne dies manuell eingeben zu müssen.

### PR DESCRIPTION
**Pull request by Bentipa**
So this closes #1590.

One “issue” that currently remains is that if you select an area (so mouse drag in a day) no tooltip is shown.
This seems to not work as Java/Swing does not recognize the "hover" anymore which is required to show the tooltip.

Apart from that, rounding was added/fixed for the creation of events in the month view to now show the correctly assumed start/end times. 
If hovering over the calendar, a tooltip shows the current selected time period (already rounded).

image
(the cursor is not visible as of screenshot mode)

@j-dimension if I should revert the auto format changes just say so, they are not related to the fix/feature

Edit: there seems still to be a bug when creating an event in month view: tooltip time != final creation time, I will further investigate

### CODE REVIEW

**Comment by Bentipa** [comment]
@j-dimension So I fixed the bug mentioned in the edit now. I additionally fixed the issue of when the mouse hovers in between 23:30 and 00:00 to show an inverted/wrong selection. Now everything should work as expected.

File: j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java [hunk 1]
```diff
@@ -133,7 +133,7 @@ public static Collection<Date> getDates(final Date start, final Date end) {
 
     public static long getTotalSeconds(final Date date) {
         final Calendar c = CalendarUtil.getCalendar(date, false);
-        long seconds = c.get(Calendar.HOUR_OF_DAY) * 60l * 60l;
+        long seconds = c.get(Calendar.HOUR_OF_DAY) * 60L * 60L;
```
Comment (j-dimension) [inline j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java:136]: just out of curiosity: does the case make a difference here?

File: j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java [hunk 2]
```diff
@@ -218,13 +218,43 @@ public static Date roundDateToHalfAnHour(Date date, boolean roundUp) {
...
-            if(roundUp)
+            if (roundUp)
+                c.add(Calendar.HOUR, 1);
+        }
+        c.set(Calendar.SECOND, 0);
+        c.set(Calendar.MILLISECOND, 0);
+        return c.getTime();
+    }
+
+    public static Date roundDateToHalfAnHour2(Date date, boolean roundUp) {
```
Comment (j-dimension) [inline j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java:231]: is this method used somewhere?

Comment (j-dimension) [reply j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java:231]: if it is supposed to remain, having  a (very brief) comment on the difference in behaviour would be beneficial

**Review by j-dimension — CHANGES_REQUESTED** [review]

File: j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java (hunk 2 above)
Comment (Bentipa) [reply j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java]: Removed, sorry for the confusion, this was accidentally added. 
I first searched for an issue in the rounding method, but then solved it by storing the selected date directly.
Previously it was Coordinates => Rounded Date => Coordinates on Move and then on Release it was Coordinates => Rounded Date again leading to different results.

File: j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java (hunk 1 above)
Comment (Bentipa) [reply j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java:136]: My IDE recommended changing it to the uppercase L :D
Seems to be a [readability recommendation](https://stackoverflow.com/questions/40953524/java-is-there-a-difference-between-l-and-l-lowercase-l-when-specifying-a-long#:~:text=No%20practical%20difference.,the%20corresponding%20Long%20wrapper%20type.&text=An%20integer%20literal%20is%20of,of%20type%20int%20(%C2%A74.2.)

Comment (j-dimension) [reply j-lawyer-client/src/de/costache/calendar/util/CalendarUtil.java:136]: Makes perfect sense :-)

Going to merge this in a few minutes

**Comment by j-dimension** [comment]
@Bentipa looking at this again I see that you added two class members startDate and endDate, and in some of the methods there are local variables of the same name. Any objections with me removing the class-level members? Does any of the changes in the PR depend on state being preserved between two method calls?

**Comment by j-dimension** [comment]
Ah, I see the mouseReleased relies on those members. Looking into this...

**Comment by j-dimension** [comment]
@Bentipa Makes sense, thanks! There is this part that can easily be confused because it creates local vars with the same name. Looking at the code, I think it is safe to just the class members all the way. Do you agree?

```
                    if (holidayEvents.isEmpty()) {
                        Date startDate = CalendarUtil.pixelToDate(
                                owner.getDate(), e.getY(),
                                getHeight());
                        Date endDate = CalendarUtil.pixelToDate(owner.getDate(),
                                e.getY(), getHeight());
                        startDate = CalendarUtil.roundDateToHalfAnHour(startDate, false);
                        endDate = CalendarUtil.roundDateToHalfAnHour(endDate, true);
                        setToolTipText(sdf.format(startDate) + " - " + sdf.format(endDate));
                    } else {
                        setToolTipText(calendar.getTooltipFormater().format(holidayEvents));
                    }
```

**Comment by j-dimension** [comment]
Same goes for this part:

```
            public void mouseClicked(final MouseEvent e) {
                for (final MouseListener ml : DayContentPanel.this.owner
                        .getOwner().getMouseListeners()) {
                    ml.mouseClicked(e);
                }
                if (e.getClickCount() == 2 && e.getButton() == MouseEvent.BUTTON1) {
                    if (startSelection == null || endSelection == null)
                        return;
                    Date startDate = CalendarUtil.pixelToDate(owner.getDate(), (int) startSelection.getY(), getHeight());
                    Date endDate = CalendarUtil.pixelToDate(owner.getDate(), (int) endSelection.getY(), getHeight());

                    EventRepository.get().triggerIntervalSelection(calendar,
                            startDate, endDate);
                }
            }
```

**Comment by j-dimension** [comment]
Changed, tested, lookin' all good! Thanks!

**Comment by Bentipa** [comment]
For me the mouseClicked listener was not event triggered once, I guess it could be removed completely.
For the other location I removed the whole date handling there, it's not required anymore.
Maybe we could keep the PR open next time until all details are clear? Then it's a little simpler to push changes afterwards :)

**Comment by j-dimension** [comment]
> Maybe we could keep the PR open next time until all details are clear? Then it's a little simpler to push changes afterwards :)

Absolutely. The code looked totally fine to me, but then after the merge Sonarqube found the duplicate variables and sent me a notification. I need to figure out how to enable it for PRs as well.

**Comment by Bentipa** [comment]
Ah okay I get it :D Yeah that would be awesome to have sonarqube feedback or something like that directly in the Review. 
Maybe a Linter could have helped as well, I will check it next time before pushing :)

**Pull request merged**